### PR TITLE
[ squash merge ] adds support for aqlm

### PR DESCRIPTION
**Pull request by shwu-nyunai**
adds support for aqlm + pv tuning

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -10,3 +10,7 @@
 [submodule "engines/mit_han_lab_qserve/QServe"]
 	path = engines/mit_han_lab_qserve/QServe
 	url = https://github.com/nyunAI/qserve.git
+[submodule "quantization/aqlm/AQLM"]
+	path = quantization/aqlm/AQLM
+	url = https://github.com/nyunAI/AQLM.git
+	branch = pv-tuning
```
Comment (Arnav0400) [inline .gitmodules:16]: can we also fix the commit ID of the submodules, since all these libraries are updated frequently?

File: core/dataset.py [hunk 2]
```diff
@@ -217,3 +217,59 @@ def get_flap_dataloader(self, nsamples=128, seed=0, seqlen=2048, tokenizer=None)
...
+        import random
+        from tqdm import trange
+
+        assert (
+            tokenizer is not None
+        ), "Tokenizer must be provided for AQLM caliberation dataloader"
+        ds = load_from_disk(self.dataset_name_or_path)
+        traindata = ds[self.split]
+        tokenizer.bos_token_id = 1
+        tokenizer.eos_token_id = 2
```
Comment (Arnav0400) [inline core/dataset.py:256]: can we infer this using the tokenizer as every model has a different token

File: quantization/aqlm/config.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    seed: int = 0
+    skip_out_loss: bool = False
+    steps_per_epoch: int = 100
+    true_sequential: bool = False
+    trust_remote_code: bool = True
+    use_checkpointing: bool = False
+    use_faiss: bool = False
+    use_fast_tokenizer: bool = False
+    val_size: int = 256
+    wandb: bool = False
```
Comment (Arnav0400) [inline quantization/aqlm/config.py:64]: I am not sure if we should have a wandb integration, if yes how is the login handled, does this conflict with our logger?

File: quantization/aqlm/config.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+                self.val_size < self.nsamples
+            ), "Number of validation set must be smaller than train + val"
+
+
+@dataclass
+class FineTuneConfig:
+
+    ### model
+    base_model: str = "base_model"
+    block_type: str = "LlamaDecoderLayer"
```
Comment (Arnav0400) [inline quantization/aqlm/config.py:90]: do we only support llama arch?

File: requirements.txt
Comment (Arnav0400) [inline requirements.txt:1]: why are the other requirements commented out?

File: quantization/aqlm/utils/tokenize_dataset.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    rank = torch.distributed.get_rank()
+    world_size = torch.distributed.get_world_size()
+    device = torch.device(f"cuda:{rank}")
+    torch.cuda.set_device(device)
+
+    args = config.finetune_config
+
+    tokenizer = transformers.AutoTokenizer.from_pretrained(args.base_model)
+    assert tokenizer.eos_token_id is not None
+    tokenizer.pad_token = tokenizer.eos_token
```
Comment (Arnav0400) [inline quantization/aqlm/utils/tokenize_dataset.py:38]: this is true only for llama models, can we add a condition to only do this when pad token is None

File: quantization/aqlm/utils/finetune.py [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+
+    if args.wandb and rank == 0:
+        assert has_wandb, "`wandb` not installed, try pip install `wandb`"
+        wandb.init(
+            config={a: getattr(args, a) for a in dir(args) if not a.startswith("_")}
+        )
+
+    tokenizer = transformers.AutoTokenizer.from_pretrained(args.base_model)
+    assert tokenizer.eos_token_id is not None
+    tokenizer.pad_token = tokenizer.eos_token
```
Comment (Arnav0400) [inline quantization/aqlm/utils/finetune.py:106]: add conditional

**Review by Arnav0400 — CHANGES_REQUESTED** [review]

File: .gitmodules (hunk 1 above)
Comment (shwu-nyunai) [reply .gitmodules:16]: the commit ids are fixed with the git hash so it always refers to the commit id with which it was added as a submodule. same for all submodules.

File: quantization/aqlm/main.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        self.config.finetune_config.dataset_name = dataset_outputs
+        self.config.finetune_config.save = str(finetune_outputs)
+
+        # conversion config
+        self.config.conversion_config.base_model = str(job.model.model_path)
+        self.config.conversion_config.quantized_model = str(job.model.model_path)
+        self.config.conversion_config.pv_fsdp_dir = self.config.finetune_config.save
+        self.config.conversion_config.save = conversion_outputs
+
+        # logger.info(f"Created AQLM instance with config: {self.config}")
```
Comment (Akash-guna) [inline quantization/aqlm/main.py:66]: This comment maybe removed

File: quantization/aqlm/config.py [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+
+@dataclass
+class AQLMConfig:
+    calibration_config: CalibrationConfig = field(default_factory=CalibrationConfig)
+    finetune_config: FineTuneConfig = field(default_factory=FineTuneConfig)
+    conversion_config: ConversionConfig = field(default_factory=ConversionConfig)
+    save_intermediate_results: bool = False
+
+
+# if __name__ == "__main__":
```
Comment (Akash-guna) [inline quantization/aqlm/config.py:214]: Are these Commented lines important

File: quantization/aqlm/config.py (hunk 3 above)
Comment (shwu-nyunai) [reply quantization/aqlm/config.py:64]: > how is the login handled

haven't tested it out but as long as wandb is logged-in 
- globally for local runtimes
- and the env variables are set in .env file in the nyuncli workspace (which gets mounted to the docker [as here](https://github.com/nyunAI/nyunzero-cli/blob/03f45562a5b80db3be8f8c4623e7bbcdf9f30831/zero/core/utils.py#L262) )
it should be working

At least that's what was happening in our experimentation stage. Also, wandb's logger is only triggered if it's installed. 

> does this conflict with our logger?

in my testing the logs were as expected
we can, for the time being, remove wandb from the recipes.

let me know your opinion.

**Review by Akash-guna — CHANGES_REQUESTED** [review]

File: quantization/aqlm/config.py (hunk 3 above)
Comment (Arnav0400) [reply quantization/aqlm/config.py:64]: lets set wandb by default to false for now, we can pick it later

File: quantization/aqlm/config.py (hunk 4 above)
Comment (shwu-nyunai) [reply quantization/aqlm/config.py:90]: no, I think this stayed from the defaults in the parsers.
Llama and Phi3 works ootb ✅

File: core/dataset.py (hunk 2 above)
Comment (shwu-nyunai) [reply core/dataset.py:256]: added a check to update only if not already set in the tokenizer.

File: quantization/aqlm/main.py (hunk 7 above)
Comment (shwu-nyunai) [reply quantization/aqlm/main.py]: cleaned in the fix commit.

File: quantization/aqlm/config.py (hunk 8 above)
Comment (shwu-nyunai) [reply quantization/aqlm/config.py]: no. forgot to clean.

File: quantization/aqlm/utils/finetune.py (hunk 6 above)
Comment (shwu-nyunai) [reply quantization/aqlm/utils/finetune.py]: fixed in the revision commit.

File: quantization/aqlm/utils/tokenize_dataset.py (hunk 5 above)
Comment (shwu-nyunai) [reply quantization/aqlm/utils/tokenize_dataset.py]: fixed in the revision commit.

File: requirements.txt
Comment (shwu-nyunai) [reply requirements.txt:1]: uncommented.

**Review by Akash-guna — APPROVED** [review]

File: quantization/aqlm/config.py (hunk 4 above)
Comment (Arnav0400) [reply quantization/aqlm/config.py:90]: is there any ways this argument can be updated via config?

Comment (Arnav0400) [reply quantization/aqlm/config.py:90]: @shwu-nyunai

Comment (shwu-nyunai) [reply quantization/aqlm/config.py:90]: yes; block_type is updated from the config.

**Pull request merged**